libbeat/processors/actions/copy_fields.go

```diff
@@ -0,0 +1,106 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package actions
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+)
+
+type copyFields struct {
+	config copyFieldsConfig
+}
+
+type copyFieldsConfig struct {
+	Fields        []fromTo `config:"fields"`
+	IgnoreMissing bool     `config:"ignore_missing"`
+	FailOnError   bool     `config:"fail_on_error"`
+}
+
+func init() {
+	processors.RegisterPlugin("copy_fields",
+		configChecked(newCopyFields,
+			requireFields("fields"),
+		),
+	)
+}
+
+func newCopyFields(c *common.Config) (processors.Processor, error) {
+	config := copyFieldsConfig{
+		IgnoreMissing: false,
+		FailOnError:   true,
+	}
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unpack the configuration of copy processor: %s", err)
+	}
+
+	f := &copyFields{
+		config: config,
+	}
+	return f, nil
+}
+
+func (f *copyFields) Run(event *beat.Event) (*beat.Event, error) {
+	var backup common.MapStr
+	if f.config.FailOnError {
+		backup = event.Fields.Clone()
+	}
+
+	for _, field := range f.config.Fields {
+		err := f.copyField(field.From, field.To, event.Fields)
+		if err != nil && f.config.FailOnError {
+			logp.Debug("copy_fields", "Failed to copy fields: %s", err)
+			event.Fields = backup
+			return event, err
+		}
+	}
+
+	return event, nil
+}
+
+func (f *copyFields) copyField(from string, to string, fields common.MapStr) error {
+	exists, _ := fields.HasKey(to)
```

Comment:
One thing that came to my mind when seeing `HasKey` here is that when I looked recently at the code I was wondering if `HasKey` works correctly in the case we have an event like this:

```
{ "a.b": {"c": "foo"}}
```
And we use `HasKey("a.b.foo")`. We should probably double check. I'm pretty sure for the mappings.Fields it does not work but it probably has a different implementation.

Reply:
It depends on what you are looking for. If you run `HasKey("a.b")`, it finds the submap due to the fast path in `mapFind`. However, if you run `HasKey("a.b.c")` (I assume that's what you have meant by `"a.b.foo"`), it cannot find the key, because it splits the key at every dot and considers the elements subkeys.
This is a compromise to speed up finding keys in a `MapStr`. At least it is consistent in all Beats. :)

Reply:
👍 I think we cover the different cases with the enhancement requests to the tests. Main thing that is important for me is that we test also the cases that don't work so in case we change behaviour in the code of `HasKey` for example, we know it has an impact.